src/satipc.c

```diff
@@ -617,51 +618,120 @@ int satip_standby_device(adapter *ad) {
+        iov[2].iov_base = NULL;
+        iov[2].iov_len  = 0;
+
+        struct mmsghdr *msg = &messages[i];
+        msg->msg_hdr.msg_iov = iov;
+        msg->msg_hdr.msg_iovlen = 2;
     }
-    if (!ad)
-        ad = get_adapter(ss->sid);
-
-    // Workaround for some poor SAT>IP servers (ex. XORO).
-    //      Some servers send KEEP-ALIVE RTP packets without valid TS payload.
-    //      Then RTP packets received without valid TS data are discarded.
-    uint8_t *b = buf;
-    if (*rb > 12 && *rb < DVB_FRAME - 12 && b[0] != 0x47) {
-        LOGM("discarding RTP packet without valid TS payload (sock %d, "
-             "socket_id "
-             "%d) [len=%d]",
-             socket, ss->id, *rb - 12);
-        *rb = 0;
-        return 1;
+
+    *rb = 0;
+
+    if (size_msg <= 0) {
+        LOG("satipc_read: Error allocating memory!")
+        return 0;
     }
 
-    if (ad && sip->ignore_packets) {
-        *rb = 0;
-        return 1;
+    rr = recvmmsg(socket, messages, num_msg, MSG_DONTWAIT, NULL);
+    if (rr <= 0) {
+        DEBUGM("satipc_read: read error or zero datagrams")
+        return 0;
     }
-    *rb -= 12;
-    return (*rb >= 0);
+
+    size_msg -= (num_msg - rr) * 1316;  // update the theoretical total size of the read buffer
+
+    // Loop over all datagrams received
+    ad = get_adapter(ss->sid);
+    uint8_t *bf1 = buf1;
+    for (i = rr - 1; i >= 0; i--) {  // loop in reverse order (i = 0; i < rr; i++)
+        struct mmsghdr *msg = &messages[i];
+        struct iovec *iov = &iovs[i * 2];
+        int dlen = 0;
+        
+        if (msg->msg_len > 0) {
+            sip->rcvp++;
+
+            copy16r(seq, bf1, 2);
+            if (sip->rtp_seq == 0xFFFF)
+                sip->rtp_seq = seq;
+            if (seq > sip->rtp_seq)
+                sip->rtp_miss++;
+            else if (seq < sip->rtp_seq)
+                sip->rtp_ooo++;
+            sip->rtp_seq = (seq + 1) & 0xFFFF;
+
+            dlen = msg->msg_len - 12;
+
+            // Free incomplete or erroneous reads
+            if (dlen < 0 || dlen > 1316)
+                dlen = 0;
+
+            // Workaround for some poor SAT>IP servers (ex. XORO).
+            //      Some servers send KEEP-ALIVE RTP packets without valid TS payload.
+            //      Then RTP packets received without valid TS data are discarded.
+            uint8_t *b = iov[1].iov_base;
+            if (dlen > 0 && dlen < DVB_FRAME && b[0] != 0x47) {
+                LOGM("discarding RTP packet without valid TS payload (sock %d, "
+                     "socket_id "
+                     "%d) [len=%d]",
+                     socket, ss->id, dlen);
+                force_ok = 1;
+                dlen = 0;
+            }
+
+            if (ad && sip->ignore_packets) {
+                force_ok = 1;
+                dlen = 0;
+            }
+
+        }
+
+        // Clear empty unused buffer space
+        if (dlen < iov[1].iov_len) {
+            int ss = (188 * (dlen / 188));
```

Comment:
If you read a packet of 12 bytes header + 188 bytes payload, this check will be false, which means the buffer will still contain holes. I would suggest doing a memset over the *rb as the cost is not big.

And of course fix the if condition.

Reply:
Sorry? This condition is right!
- `dlen` is equal to the PAYLOAD data, so the 12 bytes of the RTP header are out! See the line `dlen = msg->msg_len - 12;`
- `iov[1].iov_len` is equal to the space reserved for the PAYLOAD in the read buffer (in fact 1316 bytes).

The objective of this code is to remove the end holes of the slice if the input packets contain between 1-6 TS packets (less than the total of 7 TS packets). The strategy is then to remove the hole copying (aka moving) the data of the next slices over the last valid packet. My first implementation has to convert the free space into NULL packets, but this will interfere with the stream. So even it works I prefer to not use this approach. Now your suggestion is to memset this free space to `0x0` values? In this case the read buffer will be wasted with invalid data. Are you sure this will not generate troubles in upper layers? Futhermore, the wasted space can be important. For example, if you read 25 consecutive RTP packets with just 1 TS packet, then you're occuping 1316\*25=32900 (~32KB) for only 188\*25=4700 (~4.5KB) of valid data. You want this?

Reply:
Hi @catalinii ,

Using the technique of `memset()` the holes (fast and simple; but wasted space) the upper layer requires a simple modification. This line: https://github.com/catalinii/minisatip/blob/7ba80e07884466dfa63d97b50396b336eb9372af/src/stream.c#L909
requires a small change: For example to print only if the value is different from one indicating "internal NULL data".

So you agree with the idea of memsetting to 0xFF to indicate the "internal NULL stuff"? Then the previous line in the `stream.c` will continue discarding the TS packet but it doesn't print the error (as it's not an error).

You agree?

Reply:
Hi @catalinii ,

I'll implement this strategy:
- Process in forward order.
- Mark holes internally.
- After the end of the function I'll remove the holes (if exist) before returning.

I feel that's the best solution.
You agree?

Reply:
Hi @catalinii ,

The new implementation uses the three steps commented: process in forward order, mark holes, post-process holes.